STYLE.rst

```diff
@@ -0,0 +1,515 @@
+and use ``snake_case_t`` or ``snake_case_type``, e.g., ``size_type``
+
+.. code-block:: cpp
+
+  class MyPowerHouse
+  {
+
+  public:
+
+      void do_something();
+
+  private:
+
+      void help_something();
+
+      int32_t m_serial_number;
+
+  }; /* end class MyPowerHouse */
+
+  struct PureData
+  {
+
+      // Member data names in POD are usually short for easy access.
+      int32_t serial;
+      double x, y;
+
+  }; /* end struct PureData */
+
+In a Python class, public attributes and methods (member functions) use normal
+``snake_case``. Non-public (nothing is really private in Python) attributes and
+methods use ``_leading_underscore_snake_case`` (unmangled) and
+``__double_leading_underscore_snake_case`` (mangled).
+
+Python exceptions are Python classes and use ``CamelCase``.
+
+Do the best to name a function like ``verb_objective()`` (in both C++ and
+Python).
+
+.. code-block:: python
+
+  # function.
+  take_some_action(from_this, with_that)
+  # method.
+  some_object.do_something(with_some_information)
+
+Acronym
+----
+
+Treat acronyms like a word. Do not make them all-upper-cases in names.
+
+.. code-block:: cpp
+
+  // "Http" is treated like a word in CamelCase.
+  class HttpRequest
+  {
+      // "http" is treated like a word in snake_case.
+      void update_http_header();
+  } /* end class HttpRequest */
+
+Qt
+----
+
+For Qt sub-classes, follow the Qt naming style, but prefix with ``R`` instead
+of ``Q`` and put them in the ``modmesh`` namespace. (Why "``R``"? It is the
+next character than "``Q``" and we want to distinguish the classes derived in
+modmesh.) Use ``camelCase`` (note the leading lower-case character) for
+functions. Member data should use ``m_snake_case`` as other modmesh C++ class.
+
+Iterating Counter
+----
+
+Iterating counters start with ``i``, ``j``, ``k``.
+
+- Trivial indexing variables can be named as ``it``, ``jt``, or ``kt``.
+- Standalone ``i``, ``j``, and ``k`` should never be used to name a variable
+  because they are too short.
+
+Shorthands for Unstructured Meshes
+----
+
+Code for the unstructured meshes carries geometrical terms and needs shorthands
+to keep the line width reasonable.
+
+- Two-character names for nodes, faces, and cells:
+
+  - ``nd``: node/vertex.
+  - ``fc``: face.
+  - ``cl``: cell.
+- For example, ``icl`` is a counter of cell.
+- The following prefices often (but not always) means serial numbers:
+
+  - ``nxx``: number of ``xx``, e.g., ``ncl`` is number of cells.
+  - ``mxx``: maximum number of ``xx``, e.g., ``mfc`` is the maximum number of
+    faces.
+
+More examples:
+
+- ``clnnd`` means number of nodes belonging to a cell.
+- ``FCMND`` means maximum number of nodes for a face.
+- ``icl`` means the first-level (iterating) index of cell.
+- ``jfc`` means the second-level (iterating) index of face.
+- Some special iterators used in code, such as:
+
+  - ``clfcs(icl, ifl)``: get the ``ifl``-th face in ``icl``-th cell.
```

Comment:
Personally, I think these shorthands are hard to read.

Reply:
They are hard to read, but it is better than making the code so lengthy like:

```cpp
psolution_derivative = alg->solution_derivative + jcl*NEQ*NDIM;
for (ieq=0; ieq<NEQ; ieq++)
{
    solution_of_sub_face[ieq] = qdt * pj_solution_temporal[ieq];
    solution_of_sub_face[ieq] += (p_sub_face_center[0][0]-pj_cell_center[0]) * psolution_derivative[0];
    solution_of_sub_face[ieq] += (p_sub_face_center[0][1]-pj_cell_center[1]) * psolution_derivative[1];
}
```

I tried it long-time ago and it made it so hard to debug.  The shorthands look weird initially but they became much nicer when the developer gets used to the numerical method, which requires some weeks of derivation of the equations.

Reply:
Different modules may use different shorthands.  Those who are familiar with the numerical methods and algorithms should discuss what shorthands to use and document in the style guide.

It is about scoping. No one knows everything. Productivity comes with keeping a proper scope for development and maintenance.

Reply:
I see.